File: src/adapters/sub2api.ts
import { AppError } from '../lib/errors.js';
import type { SafeHttpClient } from '../security/safe-http.js';
import {
  bearerHeaders,
  isRecord,
  normalizePublishedAt,
  readResponseMessage,
} from './shared.js';
import type {
  AdapterAnnouncement,
  AdapterAnnouncementResult,
  AdapterSiteContext,
  SiteAdapter,
} from './types.js';

function readAnnouncementItems(value: unknown): unknown[] {
  if (Array.isArray(value)) return value;
  if (!isRecord(value)) return [];
  return Array.isArray(value.items) ? value.items : [];
}

export class Sub2ApiAdapter implements SiteAdapter {
  readonly name = 'sub2api' as const;
  readonly displayName = 'Sub2API';
  readonly capabilities = {
    checkin: false,
    announcements: true,
    requiresUserId: false,
  } as const;

  constructor(private readonly http: SafeHttpClient) {}

  async fetchAnnouncements(site: AdapterSiteContext): Promise<AdapterAnnouncementResult> {
    // Source: qixing-jk/all-api-hub and Wei-Shaw/sub2api use this JWT-protected route.
    const response = await this.http.requestJson({
      baseUrl: site.baseUrl,
      path: '/api/v1/announcements?unread_only=1',
      headers: bearerHeaders(site.accessToken),
    });
    const payload = isRecord(response.json) ? response.json : {};
    if (!response.ok || (payload.code !== undefined && payload.code !== 0)) {
      throw new AppError(
        502,
        'UPSTREAM_REJECTED',
        `Unable to fetch Sub2API announcements: ${readResponseMessage(response)}`,
        response.status >= 500,
      );
    }

    const items: AdapterAnnouncement[] = [];
    for (const candidate of readAnnouncementItems(payload.data).slice(0, 100)) {
      if (!isRecord(candidate)) continue;
      const content = [candidate.content, candidate.message, candidate.body]
        .find((value) => typeof value === 'string' && value.trim());
      const title = typeof candidate.title === 'string' ? candidate.title.trim() : '';
      if (typeof content !== 'string' && !title) continue;
      items.push({
        source: 'notice',
        content: typeof content === 'string' ? content.trim() : title,
        kind: 'account',
        extra: title && title !== content ? title : null,
        publishedAt: normalizePublishedAt(candidate.updated_at ?? candidate.created_at),
      });
    }
    return { items, warnings: [] };
  }
}

File: server/internal/adapter/sub2api.go
package adapter

import (
	"context"
	"strings"

	"github.com/elykia/apihub/server/internal/apperror"
	"github.com/elykia/apihub/server/internal/domain"
	"github.com/elykia/apihub/server/internal/netclient"
)

type Sub2APIAdapter struct{ http *netclient.Client }

func NewSub2API(http *netclient.Client) *Sub2APIAdapter { return &Sub2APIAdapter{http: http} }
func (a *Sub2APIAdapter) Descriptor() domain.AdapterDescriptor {
	return domain.AdapterDescriptor{Name: domain.Sub2API, DisplayName: "Sub2API", Capabilities: domain.Capabilities{Announcements: true}}
}
func (a *Sub2APIAdapter) CheckIn(context.Context, domain.SiteContext, string) (domain.CheckinResult, error) {
	return domain.CheckinResult{}, unsupported("Sub2API", "server-side check-in")
}
func (a *Sub2APIAdapter) FetchAnnouncements(ctx context.Context, site domain.SiteContext) (domain.AnnouncementResult, error) {
	response, err := a.http.RequestJSON(ctx, site.BaseURL, "/api/v1/announcements?unread_only=1", "GET", bearerHeaders(site.AccessToken), nil)
	if err != nil {
		return domain.AnnouncementResult{}, err
	}
	payload, _ := record(response.JSON)
	if !response.OK || (has(payload, "code") && !numberEquals(payload["code"], 0)) {
		return domain.AnnouncementResult{}, apperror.New(502, apperror.UpstreamRejected, "Unable to fetch Sub2API announcements: "+responseMessage(response), response.Status >= 500)
	}
	var candidates []any
	switch data := payload["data"].(type) {
	case []any:
		candidates = data
	case map[string]any:
		candidates, _ = data["items"].([]any)
	}
	result := domain.AnnouncementResult{}
	for index, candidate := range candidates {
		if index >= 100 {
			break
		}
		item, ok := record(candidate)
		if !ok {
			continue
		}
		content := ""
		for _, key := range []string{"content", "message", "body"} {
			if content = stringValue(item[key]); content != "" {
				break
			}
		}
		title := stringValue(item["title"])
		if content == "" {
			content = title
		}
		if content == "" {
			continue
		}
		var extra *string
		if title != "" && title != content {
			value := strings.TrimSpace(title)
			extra = &value
		}
		published := item["updated_at"]
		if published == nil {
			published = item["created_at"]
		}
		result.Items = append(result.Items, domain.AnnouncementItem{Source: "notice", Content: content, Kind: "account", Extra: extra, PublishedAt: normalizePublishedAt(published)})
	}
	return result, nil
}
